crates/aptos-crypto/src/unit_tests/p256_test.rs

```diff
@@ -0,0 +1,204 @@
+        prop_assert!(deserialized.verify_arbitrary_msg(&msg, &keypair.public_key).is_ok());
+    }
+
+    #[test]
+    fn test_signature_verification_from_struct(
+        x in any::<usize>(),
+        keypair in uniform_keypair_strategy::<P256PrivateKey, P256PublicKey>()
+    ) {
+        let hashable = CryptoHashable(x);
+        let signature = keypair.private_key.sign(&hashable).unwrap();
+        let serialized: &[u8] = &(signature.to_bytes());
+        prop_assert_eq!(P256_SIGNATURE_LENGTH, serialized.len());
+        let deserialized = P256Signature::try_from(serialized).unwrap();
+        prop_assert!(deserialized.verify(&hashable, &keypair.public_key).is_ok());
+    }
+
+
+    // Check for canonical S.
+    #[test]
+    fn test_signature_malleability(
+        message in random_serializable_struct(),
+        keypair in uniform_keypair_strategy::<P256PrivateKey, P256PublicKey>()
+    ) {
+        let signature = keypair.private_key.sign(&message).unwrap();
+        let mut serialized = signature.to_bytes();
+        let serialized_old = serialized; // implements Copy trait
+        prop_assert_eq!(serialized_old, serialized);
+
+        let mut r_bytes: [u8; 32] = [0u8; 32];
+        r_bytes.copy_from_slice(&serialized[..32]);
+
+        let mut s_bytes: [u8; 32] = [0u8; 32];
+        s_bytes.copy_from_slice(&serialized[32..]);
+
+        // NIST-P256 signing ensures a canonical S value.
+        let s = NonZeroScalar::try_from(&s_bytes[..]).unwrap();
+        let order_half = NonZeroScalar::try_from(&ORDER_HALF[..]).unwrap();
+
+        // adding ORDER_HALF (half the order of the base point) so that S + ORDER_HALF > ORDER_HALF
+        let malleable_s = NonZeroScalar::new(*s + *order_half).unwrap();
+        let malleable_s_bytes = malleable_s.to_bytes();
+        // Update the signature (the S part).
+        serialized[32..].copy_from_slice(&malleable_s_bytes);
+
+        prop_assert_ne!(serialized_old, serialized);
+
+        // Check that malleable signatures will pass verification and deserialization in the RustCrypto
+        // p256 crate.
+        // Construct the corresponding RustCrypto p256 public key.
+        let rustcrypto_public_key = p256::ecdsa::VerifyingKey::from_sec1_bytes(
+            &keypair.public_key.to_bytes()
+        ).unwrap();
+
+        // Construct the corresponding RustCrypto p256 Signature. This signature is malleable.
+        let rustcrypto_sig = p256::ecdsa::Signature::try_from(&serialized[..]);
+
+        // RustCrypto p256 will deserialize the malleable
+        // signature. It does not detect it.
+        prop_assert!(rustcrypto_sig.is_ok());
+
+        let msg_bytes = bcs::to_bytes(&message);
+        prop_assert!(msg_bytes.is_ok());
+
+        let rustcrypto_sig = rustcrypto_sig.unwrap();
+        // RustCrypto p256 verify will NOT accept the mauled signature
+        prop_assert!(rustcrypto_public_key.verify(msg_bytes.as_ref().unwrap(), &rustcrypto_sig).is_err());
```

Comment:
But if the malleability check is done in `p256` why are we redoing it in our crate?

Reply:
Fixed, the non-canonical signature was being computed incorrectly